lib/serial_terminal.pm

```diff
@@ -84,12 +84,14 @@ sub login {
     # newline nudges the guest to display the login prompt, if this behaviour
     # changes then remove it
     type_string("\n");
-    wait_serial(qr/login:\s*$/i);
+    die 'Failed to wait for login prompt' unless wait_serial(qr/login:\s*$/i);
     type_string("$user\n");
-    wait_serial(qr/Password:\s*$/i);
-    type_password;
-    type_string("\n");
-    wait_serial(qr/$escseq* \w+:~\s\# $escseq* \s*$/x);
+    if (length $testapi::password) {
+        die 'Failed to wait for password prompt' unless wait_serial(qr/Password:\s*$/i);
+        type_password;
+        type_string("\n");
+        die 'Failed to confirm that login was successful' unless wait_serial(qr/$escseq* \w+:~\s\# $escseq* \s*$/x);
```

Comment:
Does this also match for `$user` != root?

Reply:
No idea , do you know scenario where password-less non-root user trying to login ? I will clone it and check .

Reply:
That would only apply to non-root with password

Reply:
the trick that no one actually using serial terminal for non-root users , I have checked our codebase there is not a single call for `select_serial_terminal(0);`

Reply:
In that case it should be fine. It means that the $user parameter is pretty broken at the moment though.

Reply:
it is not "broken" . we just have some amount of dead code . some logic in `select_serial_terminal` which decided is it root or not . we even have function `select_user_serial_terminal` which is no one calling ...